### ISSUE TITLE
Lifecycle callbacks (onAfterEnter, onBeforeEnter, etc.) not called on page load

### ISSUE THREAD

**Issue by lpellegr** [vaadin/vaadin-router#282]
Router lifecycle callbacks are not called on the first page load. At least one page navigation is required (e.g clicking on a link, using `Router.go`, etc.) before getting lifecycle callbacks invoked for a given component.

Let's say you have for instance the following route: `/account/profile/`. You open this route on first page load. I mean directly from your browser URL  bar. You are not authenticated and the nested component associated with `/profile` defines an   `onBeforeEnter(location, commands, router)` to redirect the user if he is not authenticated (i.e. the use case described in the documentation). 

Since `onBeforeEnter` is not called, you have to perform the authentication check in one additional place (e.g. connectedCallback, only for the first invocation since subsequent will be handled by `onBeforeEnter`). If the callback was called on page load, no extra work is required and all checks would be uniformed.

Is this behavior voluntary? if yes, what's the purpose?

**Comment by lpellegr** [comment]
(Polymer 3 was used for testing)

**Comment by lpellegr** [comment]
Any update?

**Comment by vlukashov** [comment]
I've created an [online demo](https://glitch.com/edit/#!/282-lifecycle-callbacks-on-first-page-load?path=script.js:1:0) to reproduce the issue, but it does not seem to show up.  See that when opening the [`/account/profile`](https://282-lifecycle-callbacks-on-first-page-load.glitch.me/account/profile) URL in a new browser tab the [`onBeforeEnter`](https://glitch.com/edit/#!/282-lifecycle-callbacks-on-first-page-load?path=views/profile-view.js:10:4) callback does get called and redirects the user to `/login`.

This bug looks like a possible blocker for the project we are currently working on, so I would like to get a minimal reproducible example for it. So far I was not able to reproduce the issue. @lpellegr, could you please help with that? It would be great if you could modify the Glitch demo I've created so that it shows the issue.

**Comment by lpellegr** [comment]
@vlukashov Thanks for the online demo. The issue is quite old but I think I got what the problem was. Here is the adapted demo:

https://glitch.com/edit/#!/vivacious-mockingbird?path=script.js:16:6

I have defined `onBeforeEnter` in `home-view.js`:
https://glitch.com/edit/#!/vivacious-mockingbird?path=views/home-view.js:11:15

and changed `script.js` to import the component definition dynamically with import:
https://glitch.com/edit/#!/vivacious-mockingbird?path=script.js:16:1

On page load, `onBeforeEnter` from `home-view.js` is not called but it is for interactions after page load. With your demo, it is clear the problem comes from the dynamic import. In fact, the value it returns is not returned in the action callback. Once it is, the issue disappears.

To conclude, it was a usage issue and not a bug. Sorry for the noise.
